File: HackerEarth/grid.py
n,m,q = [int(x) for x in input().split()]
s = []
for i in range(n):
    s.append(input())
Q = []
lvl = 0
visited = [[0 for x in range(m+10)] for y in range(n+10)]
level   = [[0 for x in range(m+10)] for y in range(n+10)]

xy = [int(x) for x in input().split()]
Q.append((xy[0], xy[1]))
while len(Q) > 0:
    x = Q[-1]
    Q.pop()
    a, b = x[0], x[1]
    print(a, b)
    if a<0 or a > n-1 or b < 0 or b < m-1 or visited[a][b] or s[a][b] == "*":
        visited[a][b] = True
        level[a][b] = lvl

    Q.append((a+1, b))
    Q.append((a-1, b))
    Q.append((a, b+1))
    Q.append((a, b-1))

    lvl += 1

for _ in range(q):
    a, b = [int(x)-1 for x in input().split() ]
    print(level[a][b])

File: HackerEarth/grid.cpp
// Grid

#include <bits/stdc++.h>
using namespace std;

int visit[1003][1003], level[1003][1003];

string S[1003];
queue< pair<int, int> > Q;

int main()
{
cin.tie(NULL), cout.tie(NULL);

int N, M, q;
cin >> N >> M >> q;

for(int i=0; i<N; i++)
cin >> S[i];

int x, y;
cin >> x >> y;

memset(level, -1, sizeof(level));

int lvl = 0;
Q.push(make_pair(x-1, y-1));

while(!Q.empty())
{
int q = Q.size();

while(q--)
{
pair<int, int> x = Q.front();
Q.pop();
int a = x.first, b = x.second;

if(a < 0 || b < 0 || a >= N || b >= M || visit[a][b] || S[a][b] == '*')
continue;

visit[a][b] = 1;
level[a][b] = lvl;

Q.push(make_pair(a-1, b));
Q.push(make_pair(a+1, b));
Q.push(make_pair(a, b-1));
Q.push(make_pair(a, b+1));
}

lvl++;
}

while(q--)
{
int x, y;
cin >> x >> y;

cout << level[x-1][y-1] << "\n";
}

return 0;
}